### PR TITLE
[CELEBORN-1177] OpenStream should register stream via ChunkStreamManager to close stream for ReusedExchange

### PR DESCRIPTION
**Pull request by SteNicholas**
### What changes were proposed in this pull request?

`OpenStream` should register stream via `ChunkStreamManager`, which is served to obtain disk file to close stream for `ReusedExchange` operator.

Follow up #1932.

### Why are the changes needed?

`OpenStream` does not register chunk stream for reading local or dfs shuffle. Therefore `LocalPartitionReader` and `DfsPartitionReader` could not obtain the disk file from `ChunkStreamManager` that causes the below `NullPointerException` for closing stream.
```
ERROR [fetch-server-11-11] TransportRequestHandler: Error while invoking handler#receive() on RPC id 4
java.lang.NullPointerException
        at org.apache.celeborn.service.deploy.worker.storage.ChunkStreamManager.getShuffleKeyAndFileName(ChunkStreamManager.java:188)
        at org.apache.celeborn.service.deploy.worker.FetchHandler.handleEndStreamFromClient(FetchHandler.scala:344)
        at org.apache.celeborn.service.deploy.worker.FetchHandler.handleRpcRequest(FetchHandler.scala:137)
        at org.apache.celeborn.service.deploy.worker.FetchHandler.receive(FetchHandler.scala:94)
        at org.apache.celeborn.common.network.server.TransportRequestHandler.processRpcRequest(TransportRequestHandler.java:96)
        at org.apache.celeborn.common.network.server.TransportRequestHandler.handle(TransportRequestHandler.java:84)
        at org.apache.celeborn.common.network.server.TransportChannelHandler.channelRead(TransportChannelHandler.java:151)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.handler.timeout.IdleStateHandler.channelRead(IdleStateHandler.java:286)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at org.apache.celeborn.common.network.util.TransportFrameDecoder.channelRead(TransportFrameDecoder.java:74)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
        at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
        at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:745) 
```
In summary, `FetchHandler` only closes stream registered via `ChunkStreamManager`. `LocalPartitionReader` and `DfsPartitionReader` should use `ChunkStreamManager#registerStream` to close stream for deleting original unsorted disk file in `ReusedExchange` operator.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

- `FetchHandlerSuiteJ#testLocalReadSortFileOnceOriginalFileBeDeleted`
- `FetchHandlerSuiteJ#testDoNotDeleteOriginalFileWhenNonRangeLocalReadWorkInProgress`
- `ReusedExchangeSuite`

### CODE REVIEW

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala [hunk 1]
```diff
@@ -227,6 +230,7 @@ class FetchHandler(
           }
           val meta = fileInfo.getFileMeta.asInstanceOf[ReduceFileMeta]
           if (readLocalShuffle) {
+            streams.put(-1, (shuffleKey, fileName))
```
Comment (mridulm) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:233]: QQ: Is `streamId` expected to be `-1` when it is local (if yes, use that instead) ? Trying to understand how it is getting cleaned up.

Comment (SteNicholas) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:233]: @mridulm,  you could refer to `replyStreamHandler(client, rpcRequestId, -1, meta.getNumChunks, isLegacy, meta.getChunkOffsets, fileInfo.getFilePath)` for reading local shuffle, which replies with `streamId` -1. I only keep the same `streamId` when reading local shuffle.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala [hunk 2]
```diff
@@ -348,7 +353,13 @@ class FetchHandler(
...
     streamType match {
       case StreamType.ChunkStream =>
-        val (shuffleKey, fileName) = chunkStreamManager.getShuffleKeyAndFileName(streamId)
+        // LocalPartitionReader/DfsPartitionReader could not get shuffle key and disk file via stream id
+        // which does not register stream in ChunkStreamManager.
+        // Meanwhile, WorkerPartitionReader could get shuffle key and disk file from ChunkStreamManager
+        // because of stream registration of ChunkStreamManager.
+        val (shuffleKey, fileName) =
+          if (streams.containsKey(streamId)) streams.remove(streamId)
+          else chunkStreamManager.getShuffleKeyAndFileName(streamId)
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:362]: Is it better to modify chunkStreamManager.getShuffleKeyAndFileName method? For example, return null when getShuffleKeyAndFileName

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala (hunk 1 above)
Comment (RexXiong) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:233]: As readLocalShuffle may have concurrent calls, if we use -1 as the key, then the streams will ultimately only retain one result, NPE still remains when handleEndStream.

Comment (mridulm) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:233]: I have similar query as @RexXiong - thanks for elaborating it better Shuang !

**Comment by SteNicholas** [comment]
@FMX, @RexXiong, @mridulm, I have address the comments and updated the bugfix for `LocalPartitionReader` and `DfsPartitionReader`. PTAL.

**Comment by SteNicholas** [comment]
@FMX, @RexXiong, @mridulm, @cfmcgrady, I have updated the `OpenStream` behavior of `FetchHandler` that register chunk stream for original unsorted disk file of `ReusedExchange` operator. PTAL.

**Comment by cfmcgrady** [comment]
It's possible to add an end to end unit test?

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala [hunk 3]
```diff
@@ -214,49 +214,63 @@ class FetchHandler(
...
-          //    has been deleted by another range's openStream request.
-          // 2. when the current request is a range openStream request.
-          if ((endIndex != Int.MaxValue) || (endIndex == Int.MaxValue && !fileInfo.addStream(
-              streamId))) {
+          var nonRangeUnsorted = false
+          // streamId is -1 means shuffle client does not have any behavior to perform with worker, including fetching chunk, closing stream etc.
+          var streamId = -1L
+          // We must get sorted fileInfo for the following cases:
+          // 1. The current request is a range openStream request.
+          // 2. The current request is a non-range openStream, but original unsorted file has been deleted by another range's openStream request.
```
Comment (RexXiong) [inline worker/src/main/scala/org/apache/celeborn/service/deploy/worker/FetchHandler.scala:222]: IMO, addStream cannot be changed as this way, probably original files would be deleted before addStream at line 232.

**Comment by SteNicholas** [comment]
@cfmcgrady, @RexXiong, @FMX, I have added e2e test in `ReusedExchangeSuite` and addressed above comments. Meanwhile, the progresses of OpenStream, EndStream and deleting original file for reading local/dfs shuffle are same as the progresses for reading worker shuffle. PTAL.

**Review by cfmcgrady — APPROVED** [review]
LGTM, thank you for the fix.

**Comment by cfmcgrady** [comment]
thanks, merging to main(v0.5.0)/branch-0.4(v0.4.0).
